Fix: nested pre tags issue in code block

FadhlanR opened the pull request:
https://github.com/user-attachments/assets/a79570a1-1f2c-43cf-b1dd-f418f8128271


**Issue**
There was an issue where code blocks couldn't display code properly when the code included nested `<pre>` tags. The previous implementation used regex to find content to display in code blocks, but it only found the first opening `<pre>` tag and the first closing `<pre>` tag. This approach failed to handle nested `<pre>` tags correctly.

**Solution**
In this PR, I've updated the way we identify code block content by using document.createElement to find the most outer pre tag as suggested by @jurgenwerk .

FadhlanR committed:
Fix: nested pre tags issue in codeblock

FadhlanR committed:
Keep search replace block logic

FadhlanR committed:
Fix failing tests

jurgenwerk reviewed (commented):
(ignore this, I pasted my feedback in the follow up review)

jurgenwerk commented on packages/host/app/components/ai-assistant/formatted-message.gts:
```diff
@@ -249,25 +249,93 @@ class HtmlDidUpdate extends Modifier<HtmlDidUpdateSignature> {
+  let tagStack: { tag: string; startPos: number }[] = [];
+  let currentPosition = 0;
+
+  let findNextTag = (
```
This parsing is getting quite big and complicated, move from formatted message component to a html utils file

jurgenwerk commented on packages/host/app/components/ai-assistant/formatted-message.gts:
```diff
@@ -249,25 +249,93 @@ class HtmlDidUpdate extends Modifier<HtmlDidUpdateSignature> {
+          continue;
+        }
+        // Skip past the entire search/replace block
+        currentPos = endMarker + 7; // length of '>>>>>>>'
```
I am confused why you are dealing with search/replace blocks here. Parsing of the search/replace blocks is already done `extractCodeData`

jurgenwerk commented on packages/host/tests/integration/components/ai-assistant-panel/codeblocks-test.gts:
```diff
@@ -279,4 +279,103 @@ module('Integration | ai-assistant-panel | codeblocks', function (hooks) {
+      roomId,
+      '@aibot:localhost',
+      {
+        body: complexHtmlMessage,
```
Other test examples in this file seem to have `body`s in markdown, for example `body: 'This is a code snippet that I made for you\n```javascript\nconsole.log("hello world");\n```\nWhat do you think about it?',`

Is there a specific reason you are using html here? AFAIK we are using `markdownToHtml` when the message is passed to the `AiAssistantMessage` I think you can use markdown here in order to fully test the data processing flow

jurgenwerk commented on packages/host/tests/integration/components/ai-assistant-panel/codeblocks-test.gts:
```diff
@@ -279,4 +279,103 @@ module('Integration | ai-assistant-panel | codeblocks', function (hooks) {
+
+  test('handles nested pre tags', async function (assert) {
+    let roomId = await renderAiAssistantPanel(`${testRealmURL}Person/fadhlan`);
+    let complexHtmlMessage = `<p>I&#39;ll help you create an example with pre tags in HTML. Here&#39;s a useful example that demonstrates different ways to use the pre tag:</p>
```
How come content is escaped here? For example why `Here&#39;s` instead of `Here's`

Is this how the bot is sending the message in the matrix message? Does it not send it in a markdown format? We should simulate that format as closely as possible

jurgenwerk commented on packages/host/app/lib/formatted-message/utils.ts:
```diff
@@ -7,31 +7,61 @@ import {
-  if (!content) {
-    return emptyCodeData;
+  // Find the matching closing tag considering nesting
+  let closingTagStart = findMatchingClosingTag(preElementString, openTagEnd);
```
Since this is now getting a bit more complicated, did you consider an alternative of creating an element using `document.createElement`, setting the html to it and use `querySelectorAll` to get the inner parts of the 1st level `<pre>`s? Wondering if that would be easier since the browser probably already has the functionality that you want.

jurgenwerk reviewed (changes requested):
In your screen video you have escaped html code that looks like `&lt;head&gt;`..., shouldn't that be displayed with proper formatting in a human readable way? Also would be good to write a top level description of what the issue was and how you solved it. It makes reviewing easier.

FadhlanR committed:
Update based on feedback

FadhlanR committed:
Use dom manipulation for parseHtmlContent too

FadhlanR commented:
@jurgenwerk Thanks for the suggestion to use `document.createElement`. I've updated it please check it again.

FadhlanR committed:
remove console log

jurgenwerk commented on packages/host/tests/integration/components/ai-assistant-panel/codeblocks-test.gts:
```diff
@@ -279,4 +279,115 @@ module('Integration | ai-assistant-panel | codeblocks', function (hooks) {
+
+<!-- Pre tag with HTML entities -->
+<pre>
+&lt;html&gt;
```
Is this supposed to be `<html>...` and not `&lt;html&gt;...`? And similarly for the rest of the tags

FadhlanR replied:
I copied this example from ai bot response and it seems everything under the nested pre tag should be treated as text that is why the ai bot return `&lt;html&gt;` instead of `<html>`.

jurgenwerk commented on packages/host/tests/integration/components/ai-assistant-panel/codeblocks-test.gts:
```diff
@@ -279,4 +279,115 @@ module('Integration | ai-assistant-panel | codeblocks', function (hooks) {
+      .dom('.monaco-editor')
+      .exists(
+        { count: 2 },
+        'Should only have one monaco editor for the outer pre tag',
```
Count in assertion is 2 but message says "should have only one monaco editor". Which one is correct?

jurgenwerk reviewed (changes requested):
There's a bug, the diff editor is not showing diffs

Main:
<img width="273" alt="image" src="https://github.com/user-attachments/assets/73e9bd75-087d-4f61-befd-4b6462418750" />

This branch:
<img width="364" alt="image" src="https://github.com/user-attachments/assets/985c612a-58b3-401b-90f9-1bbd46d7852c" />

It's weird the tests didn't catch this, perhaps we need to improve them

FadhlanR committed:
Update test desc

FadhlanR committed:
Fix html tags outside backticks

FadhlanR committed:
Fix tests

FadhlanR committed:
Merge branch 'main' into cs-8319-fix-nested-pre-tag-in-code-blocks

FadhlanR committed:
Remove unnecessary update

FadhlanR committed:
remove console log

FadhlanR committed:
Fix reasoning tests

FadhlanR committed:
runtime lint fix

FadhlanR commented on packages/ai-bot/lib/matrix.ts:
```diff
@@ -60,12 +67,13 @@ export async function sendMessageEvent(
   commandRequests: Partial<CommandRequest>[] = [],
   reasoning: string | undefined = undefined,
 ) {
+  let html = markdownToHtml(escapeHtmlOutsideCodeBlocks(body));
```
This is another problem. The formatted message must be in html, but ai bot didn't convert the body which is in markdown into html.

FadhlanR commented on packages/host/app/components/matrix/room-message.gts:
```diff
@@ -112,6 +116,19 @@ export default class RoomMessage extends Component<Signature> {
+    ) {
+      return this.message.formattedMessage;
+    }
+    return markdownToHtml(
```
The formatted message must be already in html. However, this handle the old cases where ai bot sent formatted message in markdown.

FadhlanR committed:
lint fix

FadhlanR committed:
set global jsdom instead

FadhlanR committed:
fix host test

FadhlanR committed:
Merge branch 'main' into cs-8319-fix-nested-pre-tag-in-code-blocks

FadhlanR committed:
merge resolve

FadhlanR commented:
I fixed the bug @jurgenwerk found in this PR. It occurred because the AI bot sent the formatted message in Markdown instead of HTML. I added code to convert the message from Markdown to HTML before the AI bot sends it, and I also updated the code to handle cases where the formatted message is still in Markdown. This PR is ready for another review.

jurgenwerk committed:
Merge branch 'main' into cs-8319-fix-nested-pre-tag-in-code-blocks

jurgenwerk commented on packages/host/app/lib/formatted-message/utils.ts:
```diff
@@ -8,28 +8,26 @@ import {
 
+  let content = preElement.innerHTML;
+  // Decode HTML entities to handle special characters like < and >
+  let textarea = document.createElement('textarea');
```
Should this element be removed after we are done getting content?

jurgenwerk commented on packages/host/tests/integration/components/ai-assistant-panel/codeblocks-test.gts:
```diff
@@ -279,4 +279,291 @@ module('Integration | ai-assistant-panel | codeblocks', function (hooks) {
+      .exists({ count: 2 }, 'Should have 2 monaco editors');
+
+    let monacoContent = getMonacoContent();
+    assert.ok(
```
I think a better test would be to just assert the whole content instead of many `.includes()`

jurgenwerk commented on packages/host/tests/integration/components/ai-assistant-panel/codeblocks-test.gts:
```diff
@@ -279,4 +279,291 @@ module('Integration | ai-assistant-panel | codeblocks', function (hooks) {
+    // Check that HTML outside backticks is displayed as actual HTML
+    assert
+      .dom('[data-test-ai-message-content]')
+      .containsText(
```
Same here, I think it's a clearer and better test if you assert the whole content string, instead of many `containsText()`s

jurgenwerk reviewed (changes requested):
I am testing your branch locally and this is what I see

https://github.com/user-attachments/assets/13e6636b-704a-42ef-bab1-33fb91efa102

jurgenwerk commented on packages/host/app/lib/formatted-message/utils.ts:
```diff
@@ -110,3 +108,44 @@ export function wrapLastTextNodeInStreamingTextSpan(
+export function parseHtmlContent(htmlString: string): HtmlTagGroup[] {
+  let result: HtmlTagGroup[] = [];
+
+  let doc = document.createElement('div');
```
This one too - does this element need to be removed after working with it?

jurgenwerk replied:
I think it would be helpful for other people who will work on this to add a comment explaining why exactly we are creating an element instead of just parsing the string directly (the way it was done previously)

FadhlanR committed:
Fix and address feedback

FadhlanR committed:
lint fix

FadhlanR committed:
Merge branch 'main' into cs-8319-fix-nested-pre-tag-in-code-blocks

jurgenwerk commented:
I'm testing this locally and can reliably see bugs where the code diff is not displayed. Observe the first code block, there is no code diff:

https://github.com/user-attachments/assets/cd9dba4f-3e05-4bd7-8e1b-581de36a7548

FadhlanR committed:
Fix code diff issue

FadhlanR committed:
Merge branch 'main' into cs-8319-fix-nested-pre-tag-in-code-blocks

FadhlanR committed:
Fix code streaming

FadhlanR commented:
> I'm testing this locally and can reliably see bugs where the code diff is not displayed. Observe the first code block, there is no code diff:
> 
>  Screen.Recording.2025-04-14.at.11.06.34.mov

@jurgenwerk fixed.


https://github.com/user-attachments/assets/78dc2e95-747d-4b27-95fb-2cf4fdf122bd


https://github.com/user-attachments/assets/034124a6-07d1-49af-ace7-f11139443f4f

jurgenwerk commented:
I'm checking your 1st video and this (circled in red) is not supposed to show up in the editor. It means the search/replace block detection probably failed. 
 
<img width="510" alt="image" src="https://github.com/user-attachments/assets/cdf9deb9-f951-4053-a070-0e4fbc3adf57" />

FadhlanR commented:
> I'm checking your 1st video and this (circled in red) is not supposed to show up in the editor. It means the search/replace detection probably failed.
> 
> <img alt="image" width="510" src="https://private-user-images.githubusercontent.com/273660/433790300-cdf9deb9-f951-4053-a070-0e4fbc3adf57.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ3MTMxMjAsIm5iZiI6MTc0NDcxMjgyMCwicGF0aCI6Ii8yNzM2NjAvNDMzNzkwMzAwLWNkZjlkZWI5LWY5NTEtNDA1My1hMDcwLTBlNGZiYzNhZGY1Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxNVQxMDI3MDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ZmZlYmMyZTAyNGEyMWNjMTRiYmUxODI4MzVhYzkyMzUwMjliYmU2ZDIzNmRkOGJjOTZhOGY2YWJhMjI0ZDQ3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ILJaddAfoqzeMbS2MP_y_crBxGZs9wiKUfrZxdH3DJY">

I asked AI to generate example code with `<pre>` tag and AI response with code with random file url. It seems the code-snippet.gts file is not found, what is expected behaviour if the file is not found?

jurgenwerk commented:
@FadhlanR We don't have support for creating new files from within the assistant. There is a ticket for creating new files in this sprint. I suggest you make a video of a real use case instead that demonstrates succesful handling on nested `pre`s (something like from the repro screenshots in the ticket)

FadhlanR commented:
> @FadhlanR We don't have support for creating new files from within the assistant. There is a ticket for creating new files in this sprint. I suggest you make a video of a real use case instead that demonstrates succesful handling on nested `pre`s (something like from the repro screenshots in the ticket)

@jurgenwerk here's the screenshot for the search and replace block that includes `<pre>` tag.

https://github.com/user-attachments/assets/4e8a1cf2-42a2-44ba-824e-4d6e7872322f

FadhlanR committed:
Revert pnpm lock updates

jurgenwerk commented on packages/host/app/lib/formatted-message/utils.ts:
```diff
@@ -117,27 +119,40 @@ export interface HtmlTagGroup {
+  let result: HtmlTagGroup[] = [];
+
+  // Using dom manipulation to find pre and non-pre tags
+  // because we need to avoid the nested html tags issue.
```
```suggestion
  // We are creating a new element in the dom so that we can easily parse the content of the top level <pre> tags. Note that <pre> elements can have nested <pre> elements inside them and by querying the dom like that it's trivial to get its contents, compared to parsing the htmlString. 
```

jurgenwerk reviewed (approved):
I tested this again and didn't find any more issues. It was a wild ride but you now have pretty good understanding of how code streaming and code patching works which is awesome so thanks for going through this. I have some more comments:

1. You are using `document.createElement('div')` twice. Is it possible to only do it once? One case is for the nested pre tags. But it's not clear to me why there is another case for this. Try to consolidate if possible, or at least make a utility function for creating a new dom element end getting the content out, or at least improve the comment in the 2nd use case.

2. After merging, be on the lookout for bugs that Chris reports in the AI channel and see if this might have caused any regressions

3. Since there are now many fix/revert commits maybe squash all the commits into one or two to reduce all the noise

FadhlanR committed:
Update comments

Co-authored-by: Matic Jurglič <matic@jurglic.si>

FadhlanR committed:
Update comment

FadhlanR committed:
Merge branch 'main' into cs-8319-fix-nested-pre-tag-in-code-blocks
